### PR TITLE
Time trend for expected values

### LINKED ISSUE #477: Add time trend in risk factors
We are adding an underlying time trend in risk factors from consuming UPF (ultra-processed food).
The plan is to add time (or year) as an additional variable into risk factor equations.

- [x] The parameters need to be estimated.
- [x] It needs to be confirmed how this time trend will affect other equations in Health-GPS.
- [x] #492 
- [x] #493

### LINKED ISSUE #492: Add time trend to expected RF values
Expected values trend over time. We need to add and apply exponential time factors to the returned expected values in `RiskFactorAdjustableModel`.

### PR DESCRIPTION
**Pull request by jamesturner246**
Resolves #492 

This change adds new `StaticLinear.json` params: `ExpectedTrend`, which are raised to the power of the iteration number and used to multiply risk factor expected values, to simulate (exponentially) changing population average of nutrients over time.

These are passed into adjustable models as shared pointers rather than references -- see https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/avoid-const-or-ref-data-members.html

All adjustable RF model code is updated to use the new trends.

### CODE REVIEW

**Comment by alexdewar** [comment]
@jamesturner246 I'm guessing you meant to tag us as reviewers rather than assigning this to us?

**Comment by jamesturner246** [comment]
Oh whoops. Thanks!

File: src/HealthGPS.Input/model_parser.cpp [hunk 1]
```diff
@@ -24,29 +24,30 @@
...
+load_risk_factor_expected(const Configuration &config) {
     MEASURE_FUNCTION();
 
     const auto &info = config.modelling.baseline_adjustment;
     if (!hgps::core::case_insensitive::equals(info.format, "CSV")) {
         throw hgps::core::HgpsException{"Unsupported file format: " + info.format};
     }
 
-    auto table = hgps::RiskFactorSexAgeTable{};
+    auto table = std::make_unique<hgps::RiskFactorSexAgeTable>();
```
Comment (alexdewar) [inline src/HealthGPS.Input/model_parser.cpp:36]: Why did you change `table` to a `std::unique_ptr` btw? Not a problem; just curious.

File: src/HealthGPS.Input/model_parser.cpp [hunk 2]
```diff
@@ -268,15 +273,15 @@ load_staticlinear_risk_model_definition(const nlohmann::json &opt, const Configu
         Eigen::MatrixXd{Eigen::LLT<Eigen::MatrixXd>{policy_covariance}.matrixL()};
 
     // Risk factor expected values by sex and age.
-    hgps::RiskFactorSexAgeTable expected = load_risk_factor_expected(config);
+    std::unique_ptr<hgps::RiskFactorSexAgeTable> expected = load_risk_factor_expected(config);
```
Comment (alexdewar) [inline src/HealthGPS.Input/model_parser.cpp:276]: I'd generally use `auto` in place of longer type names:

```suggestion
    auto expected = load_risk_factor_expected(config);
```

But up to you.

File: src/HealthGPS.Input/model_parser.cpp [hunk 3]
```diff
@@ -425,10 +430,12 @@ load_ebhlm_risk_model_definition(const nlohmann::json &opt, const Configuration
     }
 
     // Risk factor expected values by sex and age.
-    hgps::RiskFactorSexAgeTable expected = load_risk_factor_expected(config);
+    std::unique_ptr<hgps::RiskFactorSexAgeTable> expected = load_risk_factor_expected(config);
```
Comment (alexdewar) [inline src/HealthGPS.Input/model_parser.cpp:433]: Ditto:

```suggestion
    auto expected = load_risk_factor_expected(config);
```

File: src/HealthGPS.Input/model_parser.cpp [hunk 4]
```diff
@@ -437,7 +444,8 @@ load_kevinhall_risk_model_definition(const nlohmann::json &opt, const Configurat
     MEASURE_FUNCTION();
 
     // Risk factor expected values by sex and age.
-    hgps::RiskFactorSexAgeTable expected = load_risk_factor_expected(config);
+    std::unique_ptr<hgps::RiskFactorSexAgeTable> expected = load_risk_factor_expected(config);
```
Comment (alexdewar) [inline src/HealthGPS.Input/model_parser.cpp:447]:
```suggestion
    auto expected = load_risk_factor_expected(config);
```

File: src/HealthGPS.Input/model_parser.cpp [hunk 5]
```diff
@@ -24,29 +24,30 @@
 
 namespace hgps::input {
 
-hgps::RiskFactorSexAgeTable load_risk_factor_expected(const Configuration &config) {
+std::unique_ptr<hgps::RiskFactorSexAgeTable>
+load_risk_factor_expected(const Configuration &config) {
```
Comment (alexdewar) [inline src/HealthGPS.Input/model_parser.cpp:28]: I suppose this *could* be a member function...

File: src/HealthGPS/risk_factor_adjustable_model.cpp [hunk 6]
```diff
@@ -31,11 +32,18 @@ struct FirstMoment {
...
+    std::shared_ptr<RiskFactorSexAgeTable> expected,
+    std::shared_ptr<std::unordered_map<core::Identifier, double>> expected_trend)
+    : expected_{std::move(expected)}, expected_trend_{std::move(expected_trend)} {}
+
+double RiskFactorAdjustableModel::get_expected(RuntimeContext &context, core::Gender sex, int age,
+                                               const core::Identifier &factor) const noexcept {
+    double expected = expected_->at(sex, factor).at(age);
+    if (expected_trend_->contains(factor)) {
+        int elapsed_time = context.time_now() - context.start_time();
+        expected *= pow(expected_trend_->at(factor), elapsed_time);
```
Comment (alexdewar) [inline src/HealthGPS/risk_factor_adjustable_model.cpp:44]: Are these arguments the wrong way round? I'm trying to wrap my head round why you'd raise to the power of `elapsed_time`.

**Review by alexdewar — APPROVED** [review]
Besides my question about the use of `pow()`, it all looks fine to me. I've made a few small suggestions.

File: src/HealthGPS/risk_factor_adjustable_model.cpp (hunk 6 above)
Comment (jamesturner246) [reply src/HealthGPS/risk_factor_adjustable_model.cpp:44]: Yeah, it' look a bit strange, but it's basically an exponential trend in time, With time factor `t`, we have `with_trend = without_trend * t^elapsed`. So for year zero, we have `with_trend = without_trend * t^0 = without_trend`, and `t` factor changes exponentially with years.

File: src/HealthGPS.Input/model_parser.cpp (hunk 1 above)
Comment (jamesturner246) [reply src/HealthGPS.Input/model_parser.cpp:36]: I noted clang-tidy began complaining about model members as const and reference complicating automatic ctor generation -- see https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/avoid-const-or-ref-data-members.htmlv and links therein.

I suppose I could have made it into a shared pointer within the model class, but thought this way we only std move a pointer to the definition ctor. :shrug:

File: src/HealthGPS.Input/model_parser.cpp (hunk 5 above)
Comment (jamesturner246) [reply src/HealthGPS.Input/model_parser.cpp:28]: Of the `AdjustableRiskFactorModelDefinition` class, you mean?

File: src/HealthGPS.Input/model_parser.cpp (hunk 1 above)
Comment (alexdewar) [reply src/HealthGPS.Input/model_parser.cpp:36]: Ah ok. That's interesting.

I think it's the right decision to return it as a `std::unique_ptr`. Out of interest, is it actually shared though? Or could it be left as a `std::unique_ptr`?

File: src/HealthGPS/risk_factor_adjustable_model.cpp (hunk 6 above)
Comment (jamesturner246) [reply src/HealthGPS/risk_factor_adjustable_model.cpp:44]: That reminds me, I need to bound these values...

File: src/HealthGPS.Input/model_parser.cpp (hunk 1 above)
Comment (jamesturner246) [reply src/HealthGPS.Input/model_parser.cpp:36]: It's shared once it enters the risk factor model definition class (with the risk factor model class). I keep it unique here and move it into a shared pointer inside the definition class. Not sure if that's good practice or not..
